### ISSUE TITLE
infinite type in intersection crashes with "Maximum call stack size exceeded"

### ISSUE THREAD

**Issue by RafaelSalguero** [Microsoft/TypeScript#16233]
**TypeScript Version:**  2.3.4
**Code**
(default `tsc --init` used)
```ts
interface TestStoreType {
    person: {
        names: {
            lastnames: {
                first: string,
                second: string;
            }
            firstname: string;
        }
        age: number;
    }
    other: string;
}

function path<T, TKey extends keyof T>(parent: string[] , key: TKey) {
    const newPath = [... parent, key];
    let x = (<TSubKey extends keyof T[TKey]>(subkey: TSubKey) => path<T[TKey], TSubKey>(newPath, subkey));
    type TResult = typeof x & { path: string[] };
    let y = x as TResult;
    return y;
}
class PathCheck<T>{
    path<TKey extends keyof T>(key: TKey) {
        return path<T, TKey>([], key);
    }
}

const x = new PathCheck<TestStoreType>();
//Should only allow correct deep property paths:
const y = x.path("person")("names")("lastnames")("first");
```

**Expected behavior:**
The program should compile and `x.path` calls should only allow deep property path names as its arguments

**Actual behavior:**
Typescript crashes with the following error:

```
RangeError: Maximum call stack size exceeded
    at instantiateType (C:\Users\rafael\AppData\Roaming\npm\node_modules\typescript\lib\tsc.js:27176:33)
    at instantiateList (C:\Users\rafael\AppData\Roaming\npm\node_modules\typescript\lib\tsc.js:26936:33)
    at instantiateTypes (C:\Users\rafael\AppData\Roaming\npm\node_modules\typescript\lib\tsc.js:26943:20)
    at instantiateTypeNoAlias (C:\Users\rafael\AppData\Roaming\npm\node_modules\typescript\lib\tsc.js:27210:44)
    at instantiateType (C:\Users\rafael\AppData\Roaming\npm\node_modules\typescript\lib\tsc.js:27184:24)
    at instantiateSignature (C:\Users\rafael\AppData\Roaming\npm\node_modules\typescript\lib\tsc.js:27050:231)
    at instantiateList (C:\Users\rafael\AppData\Roaming\npm\node_modules\typescript\lib\tsc.js:26936:33)
    at instantiateSignatures (C:\Users\rafael\AppData\Roaming\npm\node_modules\typescript\lib\tsc.js:26946:20)
    at resolveAnonymousTypeMembers (C:\Users\rafael\AppData\Roaming\npm\node_modules\typescript\lib\tsc.js:25050:38)
    at resolveStructuredTypeMembers (C:\Users\rafael\AppData\Roaming\npm\node_modules\typescript\lib\tsc.js:25185:25)
```

**Comment by sandersn** [comment]
The return type of `path` looks infinite to me. I suspect the compiler doesn't detect that.

Specifically, what is the return type of `x`?

**Comment by sandersn** [comment]
Yep, the problem is that when we start checking the return type of `PathCheck.path`, the call to `path` starts instantiating the infinite type of `x`. I can't remember what mechanisms the compiler uses to avoid this. I'll check that next.

**Comment by RafaelSalguero** [comment]
Indeed `typeof x` is an infinite type with the form:
```
<TSKey extends keyof T[TKey]>(key: TSubKey) 
=> <TSSKey extends keyof T[TKey][TSKey]>(key: TSSKey)
=> <TSSSKey extends keyof T[TKey][TSKey][TSSKey]>(key: TSSSKey)
=> <TSSSSKey extends keyof T[TKey][TSKey][TSSKey][TSSSKey]>(key: TSSSSKey)
=> ...
```

Thus x can be called for the given `TestStoreType` as:
`x("person")("names")("lastnames")("first")` 

The error does not happen if we change the line
`type TResult = typeof x & { path: string[] }`

With:
`type TResult = typeof x`

**Comment by sandersn** [comment]
Here's a much smaller repro:

```ts
function recursive() {
  let x = <T>(subkey: T) => recursive();
  return x as typeof x & { p };
}
recursive()(1)
```

I'll talk with @ahejlsberg tomorrow and see if we can find out how to avoid instantiating the infinite type.

**Comment by sandersn** [comment]
Specifically, `getIntersectionType` calls `isEmptyObjectType` which calls `resolveStructuredType`. I think this breaks some invariant in the compiler but I am not sure.

**Comment by sandersn** [comment]
Fix is up at #16335
